src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.Editor.cs

```diff
@@ -0,0 +1,255 @@
+            {
+                _renameFile = renameFile;
+                _renameType = renameType;
+                _state = state;
+                _service = service;
+                _cancellationToken = cancellationToken;
+            }
+
+            private SemanticDocument SemanticDocument => _state.SemanticDocument;
+
+            /// <summary>
+            /// operations performed by Move Type CodeAction.
+            /// </summary>
+            internal async Task<IEnumerable<CodeActionOperation>> GetOperationsAsync()
+            {
+                var solution = SemanticDocument.Document.Project.Solution;
+
+                if (_renameFile)
+                {
+                    return RenameFileToMatchTypeName(solution);
+                }
+                else if (_renameType)
+                {
+                    return await RenameTypeToMatchFileAsync(solution).ConfigureAwait(false);
+                }
+
+                return await MoveTypeToNewFileAsync(_state.TargetFileNameCandidate).ConfigureAwait(false);
+            }
+
+            /// <summary>
+            /// Given a document and a type contained in it, moves the type
+            /// out to its own document. The new document's name typically
+            /// is the type name, or is atleast based on the type name.
+            /// </summary>
+            /// <param name="documentName">name for the new document being added</param>
+            /// <remarks>
+            /// The algorithm for this, is as follows:
+            /// 1. Fork the original document that contains the type to be moved.
+            /// 2. Keep the type and required namespace containers, using statements
+            ///     and remove everything else from the forked document.
+            /// 3. Add this forked document to the solution.
+            /// 4. Finally, update the original document and remove the type from it.
+            /// </remarks>
+            private async Task<IEnumerable<CodeActionOperation>> MoveTypeToNewFileAsync(string documentName)
+            {
+                // Fork, update and add as new document.
+                var projectToBeUpdated = SemanticDocument.Document.Project;
+                var newDocumentId = DocumentId.CreateNewId(projectToBeUpdated.Id, documentName);
+
+                var solutionWithNewDocument = await AddNewDocumentWithTypeDeclarationAsync(
+                    SemanticDocument, documentName, newDocumentId, _state.TypeNode, _cancellationToken).ConfigureAwait(false);
+
+                // Get the original source document again, from the latest forked solution.
+                var sourceDocument = solutionWithNewDocument.GetDocument(SemanticDocument.Document.Id);
+
+                // update source document to add partial modifiers to type chain
+                // and/or remove type declaration from original source document.
+                var solutionWithBothDocumentsUpdated = await UpdateSourceDocumentAsync(
```

Comment:
UpdateSourceDocumentAsync is too vague.  Needs to be more like 'RemoveTypeFromDocumentAndUpdateParentChain'